extensions/amp-analytics/0.1/transport.js

```diff
@@ -118,13 +131,224 @@ export class Transport {
+   * @param {!HTMLDocument} ampDoc The AMP document
+   * @param {!Object<string,string>} transportOptions The 'transport' portion
+   * of the amp-analytics config object
+   * @param {?Function<String>=} processResponse An optional function to
+   * receive any response messages back from the cross-domain iframe
+   */
+  processCrossDomainIframe(ampDoc, transportOptions, processResponse) {
+    user().assert(transportOptions['iframe'],
+      'Cross-domain frame parameters missing');
+    const frameUrl = transportOptions['iframe'];
+    // If iframe doesn't exist for this iframe url, create it.
+    if (Transport.hasCrossDomainFrame(frameUrl)) {
+      Transport.incrementFrameUsageCount_(frameUrl);
+      this.sendExtraData(frameUrl, transportOptions['extraData']);
+    } else {
+      const frame = this.createCrossDomainIframe_(ampDoc, frameUrl,
+        transportOptions['extraData']);
+      ampDoc.body.appendChild(frame);
+    }
+    // Regardless of whether we just created it, or are re-using an existing
+    // one, send extra data if available, and wire up the response callback
+    if (processResponse) {
+      const imc = Transport.crossDomainIframes_[frameUrl].iframeMessagingClient;
+      imc.registerCallback(
+        'ampAnalyticsResponse', msg => {
+        if (msg && msg.ampAnalyticsResponse) {
+          processResponse(msg.ampAnalyticsResponse);
+        }
+      });
+    }
+  }
+
+  /**
+   * Called when a creative no longer needs its cross-domain iframe (for
+   * instance, because the creative has been removed from the DOM).
+   * Once all creatives using a frame are done with it, the frame can be
+   * destroyed.
+   * @param {!HTMLDocument} ampDoc The AMP document
+   * @param {!Object<string,string>} transportOptions The 'transport' portion
+   * of the amp-analytics config object
+   */
+  doneWithCrossDomainIframe(ampDoc, transportOptions) {
+    const frameUrl = transportOptions['iframe'];
+    if (Transport.hasCrossDomainFrame(frameUrl) &&
+      Transport.decrementFrameUsageCount_(frameUrl) <= 0) {
+      ampDoc.body.removeChild(Transport.crossDomainIframes_[frameUrl].frame);
+      delete Transport.crossDomainIframes_[frameUrl];
+    }
+  }
+
+  /**
+   * Returns whether a url of a cross-domain frame is already known
+   * @param {!string} frameUrl
+   * @return {!boolean}
+   */
+  static hasCrossDomainFrame(frameUrl) {
+    return hasOwn(Transport.crossDomainIframes_, frameUrl);
+  }
+
+  /**
+   * Sends extra data (from the amp-analytics config object) to the
+   * cross-domain iframe.
+   * @param {!string} frameUrl The URL of the frame to send the data to
+   * @param {?string=} extraData The data to send to the frame
+   */
+  sendExtraData(frameUrl, extraData) {
+    if (extraData) {
+      const frameData = Transport.crossDomainIframes_[frameUrl];
+      if (!frameData.isReady) {
+        timerFor(window).delay(() => {
+          this.sendExtraData(frameUrl, extraData);
+        }, 10);
```

Comment:
Couple of things:
1. After consideration, I do not want to use load as an indicator that the slot is ready.  Instead assume library within the frame will send a message indicating it is ready to which you will response
2.  Given #1, you should have sendExtraData (or send any post messages) store pending messages in a queue and send them upon receiving the ready message (you should consider grouping into a single message to ensure no potential race conditions though I assume post message receiver would always get them in the order they were sent)
3.  The ready signal should trigger the queue to be flushed for all creatives waiting on the frame

Waiting for an arbitrary 10 ms seems odd...